w3af/plugins/crawl/soap.py

```diff
@@ -0,0 +1,168 @@
+import sys
+from cStringIO import StringIO
+
+import zeep
+from zeep.exceptions import XMLSyntaxError
+
+import w3af.core.data.kb.knowledge_base as kb
+from w3af.core.controllers import output_manager
+from w3af.core.controllers.plugins.crawl_plugin import CrawlPlugin
+from w3af.core.data.kb.info import Info
+from w3af.core.data.options.opt_factory import opt_factory
+from w3af.core.data.options import option_types
+from w3af.core.data.parsers.doc.url import URL
+from w3af.core.data.request.fuzzable_request import FuzzableRequest
+
+
+@contextlib.contextmanager
+def redirect_stdout(new_stdout):
+    old_stdout = sys.stdout
+    try:
+        sys.stdout = new_stdout
+        yield
+    finally:
+        sys.stdout = old_stdout
+
+
+class soap(CrawlPlugin):
+    def __init__(self):
+        super(soap, self).__init__()
+        self._is_first_run = True
+        self.service_url = ''
+        self.verbosity_level = 0
+
+    def crawl(self, fuzzable_request, debugging_id):
+        url_to_crawl = fuzzable_request.get_url().url_string
+        if self._is_first_run:
+            self._is_first_run = False
+            url_to_crawl = self.service_url
+        urls_discovered = self.parse_wsdl(url_to_crawl)
+        self._save_results(urls_discovered)
+
+    def get_options(self):
+        option_list = super(soap, self).get_options()
+
+        option = opt_factory(
+            'service_url',
+            default_value=self.service_url,
+            desc='URL pointing to WSDL specification',
+            help=(
+                "Web applications which are based on SOAP protocol exposes their"
+                "specification in WSDL format. The WSDL specification gives us info"
+                "what methods are available and how to use them. It's a road map"
+                "for SOAP client."
+            ),
+            _type=option_types.STRING,
+        )
+        option_list.add(option)
+
+        option = opt_factory(
+            'verbosity_level',
+            default_value=self.verbosity_level,
+            desc='Verbosity level',
+            help=(
+                "You can specify number as below:\n"
+                "0: No verbosity. Plugin will only detect new URLs.\n"
+                "1: Plugin will detect URLs and it will add info to knowledge base"
+                "about possible operations on given URL.\n"
+                "2: Plugin will detect URLs and dump all info from WSDL spec."
+            ),
+            _type=option_types.INT,
+        )
+        option_list.add(option)
+
+        return option_list
+
+    def set_options(self, options_list):
+        super(soap, self).set_options(options_list)
+        self.service_url = options_list['service_url'].get_value()
+        self.verbosity_level = options_list['verbosity_level'].get_value()
+
+    def get_long_desc(self):
+        return (
+            "This plugin parses SOAP according to provided WSDL specification."
+            "User should specify the WSDL URL."
```

Comment:
Add better, longer, description here.

Reply:
Not longer an issue, SOAP plugin is refactored to `wsdl_parser`